### PR TITLE
[dist, trainer] feat: add CUDA graph compile support

### PR DESCRIPTION
**Pull request by wang-zerui**
### What does this PR do?

Adds CUDA Graph friendly per-block `torch.compile` support for FSDP2 training, addressing #401.

The implementation compiles leaf FSDP target module forwards in place so module identity and FSDP2 hooks are preserved, keeps FSDP collectives outside compiled regions, enforces padded static packed lengths for the compile path, and marks training-step boundaries for PyTorch CUDA Graph Trees when available.

### Checklist Before Starting

- Search for relative PRs/issues and link here: #401
- PR title follows `[{modules}] {type}: {description}` format (enforced by [check_pr_title.yml](.github/workflows/check_pr_title.yml))
  - **Allowed modules:** `agent`, `ci`, `ckpt`, `config`, `data`, `dist`, `docker`, `docs`, `logging`, `lora`, `misc`, `model`, `omni`, `optim`, `ops`, `parallel`, `perf`, `release`, `task`, `trainer`
  - **Allowed types:** `feat`, `fix`, `refactor`, `chore`, `test`
  - Breaking changes: prepend `[BREAKING]` — e.g. `[BREAKING][parallel, model] feat: dynamic batching`

### Test

Official VeOmni GPU image:

- `PYTHONPATH=/workspace/VeOmni /app/.venv/bin/python -m pytest tests/distributed/test_torch_compile.py -q` (10 passed)
- `PATH=/app/.venv/bin:$PATH make quality`

Additional manual GPU probes on a 2x A100 SXM4 official VeOmni image: FSDP2 toy model with 3 per-layer `torch.compile(mode="reduce-overhead")` blocks, `torch.compiler.cudagraph_mark_step_begin()`, grad norm, loss allreduce, and 5 optimizer steps completed successfully (`compiled=3`). A gradient-accumulation variant with 2 micro-batches per optimizer step also completed 5 optimizer steps / 10 compiled forward-backward passes.

### API and Usage Example

New training arguments:

- `train.enable_compile=True`
- `train.compile_backend`, `train.compile_mode`, `train.compile_fullgraph`, `train.compile_dynamic`

`train.enable_compile=True` requires `train.dyn_bsz=True` and `train.pad_to_length=True` so packed inputs have static shapes for CUDA Graph capture.

### Design & Code Changes

- Add `veomni.distributed.torch_compile` helpers to select leaf target modules, compile module forwards, and mark CUDA graph step boundaries.
- Wire compile options through `TrainingArguments`, `BaseTrainer`, and FSDP2 parallelization.
- Skip compile on non-CUDA devices, DDP/non-FSDP paths, and ExtraParallel models for now.
- Document compile arguments and padded packed input requirements.
- Add unit tests for compile helper behavior and argument validation.

### Checklist Before Submitting

- Read the [Contribute Guide](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md)
- Applied pre-commit checks
- Added/updated documentation
- If `tasks/` training scripts were moved or renamed: updated `docs/` examples and verified `python3 scripts/ci/check_doc_task_paths.py` passes (also enforced by the **Check doc task paths** CI workflow)
- Added tests to CI workflow (or explained why not feasible)

### CODE REVIEW

File: veomni/distributed/torch_compile.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+import torch.nn as nn
+
+from ..utils import logging
+from ..utils.device import get_device_type
+
+
+logger = logging.get_logger(__name__)
+
+
+def select_leaf_compile_modules(target_modules: Iterable[Tuple[str, nn.Module]]) -> list[Tuple[str, nn.Module]]:
```
[the remark this answers is not in the document]
Comment (wang-zerui) [reply veomni/distributed/torch_compile.py:28]: Addressed by replacing the pairwise scan with a parent-prefix set built from target FQNs; selection is now proportional to the total FQN depth rather than target_count^2.

**Comment by wang-zerui** [comment]
Hi @Luosuu @piyifan123 @Coach257 @FoolPlayer, this PR implements the per-block `torch.compile` / CUDA Graph path discussed in #401, including the padded dynamic batching guard and CUDA Graph Trees step-boundary marker. All current checks are green, and I also resolved the automated review comments (bound-method compile cycle, leaf selection complexity, explicit FSDP2 compile args, and torch.compiler-safe tests).

Update: I tightened the config validation to require `train.dyn_bsz=True` together with `train.pad_to_length=True` whenever `train.enable_compile=True`, so the compile path now directly enforces padded packed inputs as discussed in #401. Verified again on the official VeOmni GPU image: `tests/distributed/test_torch_compile.py` now has 10 passing tests, and `make quality` passes.

One note: the early bot review summaries still mention a reference-only `cuda_graph_tools` / Megatron snapshot, but that has been removed from the current PR. The current diff is limited to the 7 implementation/test/doc files shown in the Files tab.

Could you please take a look when you have a chance?

File: veomni/arguments/arguments_types.py [hunk 2]
```diff
@@ -510,7 +510,23 @@ class TrainingArguments:
     )
     enable_compile: bool = field(
         default=False,
-        metadata={"help": "Enable torch compile."},
+        metadata={"help": "Enable per-block torch.compile for CUDA Graph friendly training."},
```
Comment (Luosuu) [inline veomni/arguments/arguments_types.py:513]: let's have a dataclass named like CompileOptions to include those fields

File: veomni/distributed/torch_parallelize.py [hunk 3]
```diff
@@ -80,6 +81,11 @@ def parallelize_model_fsdp2(
     mixed_precision: MixedPrecisionConfig = MixedPrecisionConfig(enable=True),  # noqa
     basic_modules: Optional[List[str]] = None,
     muon_expert_zero_comm: bool = False,
+    enable_compile: bool = False,
```
Comment (Luosuu) [inline veomni/distributed/torch_parallelize.py:84]: same as above, we can pass CompileConfig here

File: veomni/distributed/torch_parallelize.py [hunk 4]
```diff
@@ -192,6 +200,24 @@ def parallelize_model_fsdp2(
 
     logger.info_rank0(f"extra_parallel layer pairs: {layer_pairs}")
 
+    if enable_compile:
+        if get_device_type() != "cuda":
+            logger.warning_rank0("train.enable_compile is CUDA-only for now; skip torch.compile on this device.")
+        elif parallel_state.any_extra_parallel_enabled:
+            logger.warning_rank0(
```
Comment (Luosuu) [inline veomni/distributed/torch_parallelize.py:207]: let's raise error directly here

File: veomni/distributed/torch_parallelize.py [hunk 5]
```diff
@@ -192,6 +200,24 @@ def parallelize_model_fsdp2(
...
+            )
+        else:
+            compiled_count = compile_module_forwards(
+                select_leaf_compile_modules(target_modules),
+                backend=compile_backend,
+                mode=compile_mode,
+                fullgraph=compile_fullgraph,
+                dynamic=compile_dynamic,
+            )
+            model._veomni_compile_enabled = compiled_count > 0
```
Comment (Luosuu) [inline veomni/distributed/torch_parallelize.py:219]: we can assert compiled_count > 0 here

File: veomni/distributed/torch_parallelize.py [hunk 6]
```diff
@@ -427,6 +453,11 @@ def build_parallelize_model(
     enable_gradient_checkpointing: bool = True,
     basic_modules: Optional[List[str]] = None,
     muon_expert_zero_comm: bool = False,
+    enable_compile: bool = False,
```
Comment (Luosuu) [inline veomni/distributed/torch_parallelize.py:456]: pass CompileConfig

File: veomni/distributed/torch_parallelize.py [hunk 7]
```diff
@@ -475,9 +506,18 @@ def build_parallelize_model(
                 mixed_precision=mixed_precision,
                 basic_modules=basic_modules,
                 muon_expert_zero_comm=muon_expert_zero_comm,
+                enable_compile=enable_compile,
```
Comment (Luosuu) [inline veomni/distributed/torch_parallelize.py:509]: pass CompileConfig

File: veomni/distributed/torch_parallelize.py [hunk 8]
```diff
@@ -475,9 +506,18 @@ def build_parallelize_model(
...
+                compile_backend=compile_backend,
+                compile_mode=compile_mode,
+                compile_fullgraph=compile_fullgraph,
+                compile_dynamic=compile_dynamic,
                 **kwargs,
             )
         else:
+            if enable_compile:
+                logger.warning_rank0("train.enable_compile is only wired for fsdp2; skip compile for DDP.")
```
Comment (Luosuu) [inline veomni/distributed/torch_parallelize.py:518]: why?

File: veomni/distributed/torch_parallelize.py [hunk 9]
```diff
@@ -475,9 +506,18 @@ def build_parallelize_model(
...
             model = DDP(model, device_ids=[parallel_state.local_rank], process_group=parallel_state.dp_group)
+    elif enable_compile:
+        logger.warning_rank0("train.enable_compile is only wired for fsdp2; skip compile without FSDP.")
```
Comment (Luosuu) [inline veomni/distributed/torch_parallelize.py:521]: we have dropped fsdp1

**Review by Luosuu — COMMENTED** [review]
Thank you for your contributions! I just have some minimal comments

File: veomni/arguments/arguments_types.py [hunk 10]
```diff
@@ -1166,6 +1182,13 @@ def __post_init__(self):
                 self.train.pad_to_length = self.train.micro_batch_size * self.data.max_seq_len
                 logger.info_rank0(f"set pad_to_length = micro_batch_size * max_seq_len = {self.train.pad_to_length}")
 
+        if self.train.enable_compile and (not self.train.dyn_bsz or not self.train.pad_to_length):
```
Comment (Luosuu) [inline veomni/arguments/arguments_types.py:1185]: I think we need to limit to text trainer only as well? now we have not implemented pad_to_length for multimodal data (including both DiT and VLM)
